### PR TITLE
[WIP] Add abi.encode and abi.encodePacked

### PR DESCRIPTION
**Pull request by axic**
Part of #1707.

### CODE REVIEW

**Comment by axic** [comment]
Needs tests.

**Comment by chriseth** [comment]
Great!
We should add a test that checks that we get a warning for `abi.encode(8)`

**Comment by axic** [comment]
```
Error: Function declared as pure, but this expression (potentially) reads from the environment or state and thus requires "view".
				return abi.encodePacked(1);
				       ^--------------^
```

but `abi.encodePacked` is marked as pure.

@chriseth any ideas?

**Comment by axic** [comment]
Also `var a = abi.encode;` results in taking it as a member and crashes in
```
solidity/libsolidity/codegen/ExpressionCompiler.cpp(1139): Throw in function virtual bool dev::solidity::ExpressionCompiler::visit(const dev::solidity::MemberAccess &)
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: Unknown magic member.
[dev::tag_comment*] = Unknown magic member.
```

**Comment by axic** [comment]
It seems the encoder is off:
```
/home/travis/build/ethereum/solidity/test/libsolidity/SolidityEndToEndTest.cpp(10165): error in "abi_encode": Invalid encoded data
   Result                                                           Expectation
 X ..............................................................2. ..............................................................2.
 X ..............................................................6. ..............................................................6.
 X ................................................................ ..............................................................4.
 X ............................................................61.. ...............................................................1
 X ..............................................................2. ...............................................................2
/home/travis/build/ethereum/solidity/test/libsolidity/SolidityEndToEndTest.cpp(10166): error in "abi_encode": Invalid encoded data
   Result                                                           Expectation
 X ..............................................................2. ..............................................................2.
 X ..............................................................6. ..............................................................6.
 X ................................................................ ..............................................................4.
 X ............................................................61.. ...............................................................1
 X ..............................................................2. ...............................................................2
```

**Comment by axic** [comment]
I think this has been fixed & optimised now.

@chriseth what other tests should we do?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -897,6 +897,38 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
...
 			break;
 		}
+		case FunctionType::Kind::ABIEncode:
+		case FunctionType::Kind::ABIEncodePacked:
+		{
+			TypePointers argumentTypes;
+			for (auto const& arg: arguments)
+			{
+				arg->accept(*this);
+				argumentTypes.push_back(arg->annotation().type);
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: I think you have to use something like `.type->mobileType()`.

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: But perhaps that is done in `encodeToMemory`.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -897,6 +897,38 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
...
+			}
+			utils().fetchFreeMemoryPointer();
+			// stack now: <memory pointer>
+
+			// adjust by 32 bytes to accommodate the length
+			m_context << u256(32) << Instruction::ADD;
+			utils().encodeToMemory(argumentTypes, TypePointers(), function.padArguments(), true);
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: `copyDynamicDataInPlace` should only be true if `padArguments` is also true.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 3]
```diff
@@ -897,6 +897,38 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
...
+			utils().toSizeAfterFreeMemoryPointer();
+			// stack now: <memory size> <memory pointer>
+
+			// adjust length to create the `bytes` payload size
+			m_context << u256(32) << Instruction::DUP3 << Instruction::SUB;
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Why dup3 and not dup2?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 4]
```diff
@@ -10144,9 +10146,26 @@ BOOST_AUTO_TEST_CASE(constant_string)
...
+			}
+			function g() returns (bytes) {
+				return abi.encodePacked(uint256(1), uint256(2));
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "C");
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(u256(32), u256(64), u256(1), u256(2)));
+	ABI_CHECK(callContractFunction("g()"), encodeArgs(u256(32), u256(64), u256(1), u256(2)));
 }
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Please also add tests for dynamic data and even using the new encoder.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp]: This is a copy of the `SHA3` case above.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp]: This is a copy of the `SHA3` case above. The only difference is the `+32` for the offset.

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Yes and `sha3` does not use ABI encoding.

**Pull request merged**